Library/Homebrew/cmd/shellenv.sh

```diff
@@ -11,7 +11,7 @@
 # shellcheck disable=SC2154
 homebrew-shellenv() {
   if [[ "${HOMEBREW_SHELLENV_PREFIX}" == "${HOMEBREW_PREFIX}" ]] &&
-     [[ "$(PATH="${HOMEBREW_PATH}" command -v brew)" == "${HOMEBREW_PREFIX}/bin/brew" ]]
+     [[ "${HOMEBREW_PATH%%/usr/bin:*}" == *"${HOMEBREW_PREFIX}/bin"* ]]
```

Comment:
```suggestion
     [[ "$(PATH="${HOMEBREW_PATH}" command -v brew)" == "${HOMEBREW_PREFIX}/bin/brew" ]] &&
     [[ "${HOMEBREW_PATH%%/usr/bin:*}" == *"${HOMEBREW_PREFIX}/bin"* ]]
```

It's still worth to check the `brew` executable, because `HOMEBREW_SHELLENV_PREFIX` can be inherited to the child process. `brew shellenv` could produce empty output on multiple installation setup.

Reply:
This would matter when
```bash
[[ "$(PATH="${HOMEBREW_PATH}" command -v brew)" == "${HOMEBREW_PREFIX}/bin/brew" ]]
```
false but
```bash
[[ "${HOMEBREW_PATH%%/usr/bin:*}" == *"${HOMEBREW_PREFIX}/bin"* ]]
```
is true. Could you give an example when this is the case? It clearly only applies if you have multiple installations of `brew`, as you point out.

Reply:
Some users set `PATH` to a fixed value at the beginning of their shell profile rather than prepend or append entries to it. Then they run `eval "$(brew shellenv)"` to get the remaining paths (e.g. `MANPATH`).

Reply:
```suggestion
     [[ ":${HOMEBREW_PATH%%/usr/bin:*}:" == *":${HOMEBREW_PREFIX}/bin:"* ]]
```

Reply:
> Some users set `PATH` to a fixed value at the beginning of their shell profile rather than prepend or append entries to it. Then they run `eval "$(brew shellenv)"` to get the remaining paths (e.g. `MANPATH`).

I'm confused -- how would this make
```bash
[[ "$(PATH="${HOMEBREW_PATH}" command -v brew)" == "${HOMEBREW_PREFIX}/bin/brew" ]]
```
be false but
```bash
[[ "${HOMEBREW_PATH%%/usr/bin:*}" == *"${HOMEBREW_PREFIX}/bin"* ]]
```
be true?

The typical case of the scenario you're describing is that they're either both true or both false. There is the odd case where the first is true but the second is false, but that's for a user that's deliberately putting `/usr/bin` ahead of `${HOMEBREW_PREFIX}/bin`, so I doubt that user is calling `brew shellenv`.

As for your suggestion: I'm not sure it's correct since I don't strip `:` in front of `/usr/bin` on purpose, so the left-hand side will usually look something like `:foo:bar:baz::`. It probably won't matter because of the `*` globs on either side, but then I'm not sure what it really adds.

Reply:
In shell profile:

```bash
export PATH="/prefix2/bin:/prefix1/bin:...:/usr/bin:..."
eval "$(/prefix1/bin/brew shellenv)"
eval "$(/prefix2/bin/brew shellenv)"
```

the `PATH` is set to fixed value first, then be updated by `eval ...`.

```console
shlvl1 $ # brew in prefix2 is default and `HOMEBREW_SHELLENV_PREFIX` is prefix2

shlvl1 $ eval "$(/prefix1/bin/brew shellenv)"
shlvl1 $ # now brew in prefix1 is default and `HOMEBREW_SHELLENV_PREFIX` is prefix1

shlvl1 $ bash -l

shlvl2 $
```

in the child process, the first shellenv command will output empty result.

------

The two `:` added to the right side is to match the exact entry we want.

Reply:
How about just checking whether `$HOMEBREW_PREFIX/bin` comes first in `PATH`, as it would after doing `eval $(brew shellenv)`?

Reply:
> How about just checking whether `$HOMEBREW_PREFIX/bin` comes first in `PATH`, as it would after doing `eval $(brew shellenv)`?

If so, we will not need the variable `HOMEBREW_SHELLENV_PREFIX` anymore. It was introduced for making `eval "$(brew shellenv)"` idempotent.

```bash
eval "$(brew shellenv)"
export PATH="${HOMEBREW_PREFIX}/opt/formula/bin:${PATH}"  # add some keg-only formula to PATH

...

eval "$(brew shellenv)"  # added by `echo 'eval "$(brew shellenv)"' >> ~/.profile`
```

Reply:
Ok, let's check that it comes first in `PATH`. Not having to carry around `HOMEBREW_SHELLENV_PREFIX` is nice to have.